File: ubel-node/src/rust_runner.js
// RustCargoScanner.js
import fs   from "fs";
import path from "path";

/**
 * Scans Rust / Cargo projects for installed crates.
 *
 * Detection strategy:
 *   1.  A directory is a Cargo project root when it contains Cargo.toml
 *       AND Cargo.lock (lock file = installed state).
 *   2.  Installed packages are read from Cargo.lock (the only reliable
 *       source of fully-resolved crate graph with exact versions).
 *   3.  Workspace support: if the root Cargo.toml contains [workspace],
 *       member paths are scanned recursively.
 *   4.  PURL: pkg:cargo/<name>@<version>
 *
 * Cargo.lock format (v1/v2/v3 are all handled):
 *   [[package]]
 *   name    = "foo"
 *   version = "1.2.3"
 *   source  = "registry+..."
 *   checksum = "..."
 *   dependencies = [ "bar 0.4.0", "baz 1.0.0 (...)" ]
 */
export class RustCargoScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // ─────────────────────────────
  static _cargoPurl(name, version) {
    return `pkg:cargo/${name.toLowerCase()}@${version ?? ""}`;
  }

  // ─────────────────────────────
  // Detect Cargo project root
  // ─────────────────────────────
  static _isCargoRoot(dir) {
    return (
      fs.existsSync(path.join(dir, "Cargo.toml")) &&
      fs.existsSync(path.join(dir, "Cargo.lock"))
    );
  }

  // ─────────────────────────────
  // Minimal TOML block parser for Cargo.lock
  // Returns array of { name, version, source, dependencies[] }
  // ─────────────────────────────
  static _parseCargoLock(lockPath) {
    let content;
    try {
      content = fs.readFileSync(lockPath, "utf8");
    } catch {
      return [];
    }

    const packages = [];
    // Split on [[package]] sections
    const blocks = content.split(/^\[\[package\]\]/m).slice(1);

    for (const block of blocks) {
      const getName    = block.match(/^name\s*=\s*"([^"]+)"/m);
      const getVer     = block.match(/^version\s*=\s*"([^"]+)"/m);
      const getSrc     = block.match(/^source\s*=\s*"([^"]+)"/m);

      if (!getName || !getVer) continue;

      const name    = getName[1];
      const version = getVer[1];
      const source  = getSrc ? getSrc[1] : "local";

      // dependencies block – single-line array or multi-line array
      // e.g.:  dependencies = [\n "bar 0.4.0",\n "baz 1.0.0 (registry+...)"\n]
      const deps = [];
      const depsMatch = block.match(/^dependencies\s*=\s*\[([^\]]*)\]/ms);
      if (depsMatch) {
        const inner = depsMatch[1];
        // Each dep is a quoted string like "name version" or "name version (source)"
        const depRe = /"([^"]+)"/g;
        let dm;
        while ((dm = depRe.exec(inner)) !== null) {
          const parts = dm[1].split(" ");
          deps.push({ name: parts[0], version: parts[1] ?? "" });
        }
      }

      packages.push({ name, version, source, dependencies: deps });
    }

    return packages;
  }

  // ─────────────────────────────
  // Read [dependencies] / [dev-dependencies] / [build-dependencies]
  // from a Cargo.toml file.
  // Returns { prod: Set<string>, dev: Set<string>, build: Set<string> }
  // (all lowercase crate names)
  // ─────────────────────────────
  static _readCargoTomlDeps(tomlPath) {
    const prod  = new Set();
    const dev   = new Set();
    const build = new Set();

    let content;
    try {
      content = fs.readFileSync(tomlPath, "utf8");
    } catch {
      return { prod, dev, build };
    }

    // We do minimal section-aware line scanning (no full TOML parser needed).
    let section = "";
    for (const line of content.split("\n")) {
      const trimmed = line.trim();

      // Section header
      const secMatch = trimmed.match(/^\[([^\]]+)\]/);
      if (secMatch) {
        section = secMatch[1].trim();
        continue;
      }

      // Key = value lines inside dependency sections
      if (
        section === "dependencies"       ||
        section === "dev-dependencies"   ||
        section === "build-dependencies" ||
        // target-scoped: [target.'cfg(...)'.dependencies]
        section.endsWith(".dependencies")
      ) {
        const kvMatch = trimmed.match(/^([A-Za-z0-9_-]+)\s*[=.]/);
        if (kvMatch) {
          const name = kvMatch[1].toLowerCase().replace(/-/g, "_"); // Cargo normalises - → _
          if (section === "dev-dependencies") dev.add(name);
          else if (section === "build-dependencies") build.add(name);
          else prod.add(name);
        }
      }
    }

    return { prod, dev, build };
  }

  // ─────────────────────────────
  // Scan a single Cargo project
  // ─────────────────────────────
  static _scanProject(projectRoot) {
    const lockPath = path.join(projectRoot, "Cargo.lock");
    const packages = this._parseCargoLock(lockPath);
    if (!packages.length) return [];

    // Build lookup index: "<name>@<version>" → package entry
    const index = new Map();
    for (const pkg of packages) {
      const key = `${pkg.name.toLowerCase()}@${pkg.version}`;
      if (!index.has(key)) index.set(key, pkg);
    }

    // Also a name-only index (latest version wins if dupes) for dep resolution
    const nameIndex = new Map();
    for (const pkg of packages) {
      const norm = pkg.name.toLowerCase().replace(/-/g, "_");
      nameIndex.set(norm, pkg);
    }

    const components = [];

    for (const pkg of packages) {
      const name    = pkg.name.toLowerCase().replace(/-/g, "_");
      const id      = this._cargoPurl(pkg.name, pkg.version);
      const isLocal = pkg.source === "local" || !pkg.source.startsWith("registry");

      const dependencies = pkg.dependencies.map(dep => {
        const depName = dep.name.toLowerCase().replace(/-/g, "_");
        const depKey  = `${depName}@${dep.version}`;
        const resolved = index.get(depKey) ?? nameIndex.get(depName);
        return resolved
          ? this._cargoPurl(resolved.name, resolved.version)
          : this._cargoPurl(dep.name, dep.version);
      });

      components.push({
        id,
        name,
        version:      pkg.version,
        type:         "library",
        license:      "unknown",
        ecosystem:    "rust",
        state:        "undetermined",
        scopes:       [],
        dependencies,
        paths:        [isLocal ? projectRoot : pkg.source],
        project_root: projectRoot,
        _source:      pkg.source
      });
    }

    return components;
  }

  // ─────────────────────────────
  // Assign scopes (prod / dev / build)
  // ─────────────────────────────
  static _assignScopes(inventory) {
    const byId    = new Map(inventory.map(c => [c.id, c]));
    const nameIdx = new Map();

    for (const comp of inventory) {
      if (!Array.isArray(comp.scopes)) comp.scopes = [];
      const key   = comp.name;
      const comps = nameIdx.get(key) ?? [];
      comps.push(comp);
      nameIdx.set(key, comps);
    }

    const projectGroups = new Map();
    for (const comp of inventory) {
      const root = comp.project_root;
      if (!projectGroups.has(root)) projectGroups.set(root, []);
      projectGroups.get(root).push(comp);
    }

    for (const [projectRoot, comps] of projectGroups.entries()) {
      const { prod, dev, build } = this._readCargoTomlDeps(
        path.join(projectRoot, "Cargo.toml")
      );

      const propagate = (names, scope) => {
        const queue = [];
        for (const n of names) {
          for (const c of (nameIdx.get(n) ?? [])) {
            if (c.project_root === projectRoot) queue.push(c);
          }
        }
        const visited = new Set();
        while (queue.length) {
          const c = queue.shift();
          if (visited.has(c.id)) continue;
          visited.add(c.id);
          if (!c.scopes.includes(scope)) c.scopes.push(scope);
          for (const dep of c.dependencies) {
            const d = byId.get(dep);
            if (d && d.project_root === projectRoot) queue.push(d);
          }
        }
      };

      propagate(prod,  "prod");
      propagate(dev,   "dev");
      propagate(build, "build");

      // Fallback
      for (const c of comps) {
        if (c.scopes.length === 0) c.scopes.push("prod");
      }
    }
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }
      const existing = map.get(comp.id);
      for (const p of comp.paths)  { if (!existing.paths.includes(p))  existing.paths.push(p); }
      for (const s of comp.scopes) { if (!existing.scopes.includes(s)) existing.scopes.push(s); }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {
    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (["node_modules", ".git", ".ubel", "target"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);

        if (RustCargoScanner._isCargoRoot(full)) {
          const key = path.resolve(full);
          if (!visited.has(key)) {
            visited.add(key);
            raw.push(...RustCargoScanner._scanProject(full));
          }
          // Still descend – workspaces have nested member crates
          // BUT skip target/ which is handled above
        }

        walk(full);
      }
    }

    if (RustCargoScanner._isCargoRoot(startDir)) {
      const key = path.resolve(startDir);
      if (!visited.has(key)) {
        visited.add(key);
        raw.push(...RustCargoScanner._scanProject(startDir));
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this._assignScopes(merged);

    for (const c of merged) delete c._source;

    this.inventoryData = merged;
    return merged.map(c => c.id);
  }
}

export default RustCargoScanner;

File: python/ubel/rust_runner.py
"""
rust_scanner.py — Rust / Cargo scanner (getInstalled only).

Mirrors the JS RustCargoScanner:
  - Detects Cargo project roots (Cargo.toml + Cargo.lock)
  - Parses Cargo.lock (v1/v2/v3 [[package]] format) for fully-resolved crates
  - Reads Cargo.toml [dependencies] / [dev-dependencies] / [build-dependencies]
    for scope seeding; BFS propagates through the dep graph
  - Descends into workspace member crates
  - Merges duplicate PURLs
  - PURL: pkg:cargo/<name>@<version>
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any, Dict, List, Set


class RustCargoScanner:

    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    
    def _cargo_purl(self,name: str, version: str) -> str:
        return f"pkg:cargo/{name.lower()}@{version or ''}"

    # ------------------------------------------------------------------ #
    # Detect Cargo root                                                    #
    # ------------------------------------------------------------------ #

    
    def _is_cargo_root(self,directory: str) -> bool:
        return (
            os.path.exists(os.path.join(directory, "Cargo.toml")) and
            os.path.exists(os.path.join(directory, "Cargo.lock"))
        )

    # ------------------------------------------------------------------ #
    # Parse Cargo.lock                                                     #
    # ------------------------------------------------------------------ #

    
    def _parse_cargo_lock(self,lock_path: str) -> List[Dict[str, Any]]:
        try:
            content = Path(lock_path).read_text(encoding="utf-8")
        except OSError:
            return []

        packages: List[Dict[str, Any]] = []
        # Split on [[package]] sections
        blocks = re.split(r"^\[\[package\]\]", content, flags=re.MULTILINE)[1:]

        for block in blocks:
            name_m = re.search(r'^name\s*=\s*"([^"]+)"',    block, re.MULTILINE)
            ver_m  = re.search(r'^version\s*=\s*"([^"]+)"', block, re.MULTILINE)
            src_m  = re.search(r'^source\s*=\s*"([^"]+)"',  block, re.MULTILINE)

            if not name_m or not ver_m:
                continue

            name    = name_m.group(1)
            version = ver_m.group(1)
            source  = src_m.group(1) if src_m else "local"

            # dependencies = [ "bar 0.4.0", "baz 1.0.0 (...)" ]
            deps: List[Dict[str, str]] = []
            deps_m = re.search(
                r"^dependencies\s*=\s*\[([^\]]*)\]", block, re.MULTILINE | re.DOTALL
            )
            if deps_m:
                for dm in re.finditer(r'"([^"]+)"', deps_m.group(1)):
                    parts = dm.group(1).split()
                    deps.append({"name": parts[0], "version": parts[1] if len(parts) > 1 else ""})

            packages.append({"name": name, "version": version, "source": source, "dependencies": deps})

        return packages

    # ------------------------------------------------------------------ #
    # Read Cargo.toml dep sections                                         #
    # ------------------------------------------------------------------ #

    
    def _read_cargo_toml_deps(self,toml_path: str) -> tuple:
        """Returns (prod: Set[str], dev: Set[str], build: Set[str])"""
        prod:  Set[str] = set()
        dev:   Set[str] = set()
        build: Set[str] = set()

        try:
            content = Path(toml_path).read_text(encoding="utf-8")
        except OSError:
            return prod, dev, build

        section = ""
        for line in content.splitlines():
            trimmed = line.strip()

            sec_m = re.match(r"^\[([^\]]+)\]", trimmed)
            if sec_m:
                section = sec_m.group(1).strip()
                continue

            is_dep_section = (
                section == "dependencies" or
                section == "dev-dependencies" or
                section == "build-dependencies" or
                section.endswith(".dependencies")
            )
            if not is_dep_section:
                continue

            kv_m = re.match(r"^([A-Za-z0-9_-]+)\s*[=.]", trimmed)
            if kv_m:
                # Cargo normalises hyphens to underscores in crate names
                name = kv_m.group(1).lower().replace("-", "_")
                if section == "dev-dependencies":
                    dev.add(name)
                elif section == "build-dependencies":
                    build.add(name)
                else:
                    prod.add(name)

        return prod, dev, build

    # ------------------------------------------------------------------ #
    # Scan one Cargo project                                               #
    # ------------------------------------------------------------------ #

    
    def _scan_project(self,project_root: str) -> List[Dict[str, Any]]:
        packages = self._parse_cargo_lock(
            os.path.join(project_root, "Cargo.lock")
        )
        if not packages:
            return []

        # Build lookup indexes
        full_index: Dict[str, Any] = {}    # "name@version" → package
        name_index: Dict[str, Any] = {}    # "normalized_name" → package (last wins)

        for pkg in packages:
            key       = f"{pkg['name'].lower()}@{pkg['version']}"
            norm_name = pkg["name"].lower().replace("-", "_")
            if key not in full_index:
                full_index[key] = pkg
            name_index[norm_name] = pkg

        components: List[Dict[str, Any]] = []

        for pkg in packages:
            name    = pkg["name"].lower().replace("-", "_")
            cid     = self._cargo_purl(pkg["name"], pkg["version"])
            is_local = pkg["source"] == "local" or not pkg["source"].startswith("registry")

            dependencies: List[str] = []
            for dep in pkg["dependencies"]:
                dep_name = dep["name"].lower().replace("-", "_")
                dep_key  = f"{dep_name}@{dep['version']}"
                resolved = full_index.get(dep_key) or name_index.get(dep_name)
                if resolved:
                    dependencies.append(
                        self._cargo_purl(resolved["name"], resolved["version"])
                    )
                else:
                    dependencies.append(self._cargo_purl(dep["name"], dep["version"]))

            components.append({
                "id":           cid,
                "name":         name,
                "version":      pkg["version"],
                "type":         "library",
                "license":      "unknown",
                "ecosystem":    "rust",
                "state":        "undetermined",
                "scopes":       [],
                "dependencies": dependencies,
                "paths":        [project_root if is_local else pkg["source"]],
                "project_root": project_root,
                "_source":      pkg["source"],
            })

        return components

    # ------------------------------------------------------------------ #
    # Assign scopes                                                        #
    # ------------------------------------------------------------------ #

    
    def _assign_scopes(self,inventory: List[Dict[str, Any]]) -> None:
        by_id: Dict[str, Dict] = {c["id"]: c for c in inventory}
        name_idx: Dict[str, List[Dict]] = {}

        for comp in inventory:
            if not isinstance(comp.get("scopes"), list):
                comp["scopes"] = []
            name_idx.setdefault(comp["name"], []).append(comp)

        project_groups: Dict[str, List[Dict]] = {}
        for comp in inventory:
            project_groups.setdefault(comp["project_root"], []).append(comp)

        for project_root, comps in project_groups.items():
            prod, dev, build = self._read_cargo_toml_deps(
                os.path.join(project_root, "Cargo.toml")
            )

            def propagate(names: Set[str], scope: str) -> None:
                queue = [
                    c for n in names
                    for c in name_idx.get(n, [])
                    if c["project_root"] == project_root
                ]
                visited: Set[str] = set()
                while queue:
                    c = queue.pop(0)
                    if c["id"] in visited:
                        continue
                    visited.add(c["id"])
                    if scope not in c["scopes"]:
                        c["scopes"].append(scope)
                    for dep_id in c["dependencies"]:
                        d = by_id.get(dep_id)
                        if d and d.get("project_root") == project_root:
                            queue.append(d)

            propagate(prod,  "prod")
            propagate(dev,   "dev")
            propagate(build, "build")

            for c in comps:
                if not c["scopes"]:
                    c["scopes"].append("prod")

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    
    def merge_inventory_by_purl(self,components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        skip = {"node_modules", ".git", ".ubel", "target"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    if self._is_cargo_root(full):
                        key = os.path.realpath(full)
                        if key not in visited:
                            visited.add(key)
                            raw.extend(self._scan_project(full))
                        # Still descend – workspace member crates live inside
                    walk(full)

        if self._is_cargo_root(start_dir):
            key = os.path.realpath(start_dir)
            if key not in visited:
                visited.add(key)
                raw.extend(self._scan_project(start_dir))

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self._assign_scopes(merged)

        for c in merged:
            c.pop("_source", None)

        self.inventory_data = merged
        return [c["id"] for c in merged]